### PR TITLE
Eliminate overlaps in the Beam pipeline's leveldb shards.

### PR DESCRIPTION
**Pull request by johnedmonds**
Although LevelDB's level 0 shards are allowed to overlap, this makes initial requests extremely slow. This PR eliminates the overlaps in the written shards and improves query speed.

### CODE REVIEW

File: kythe/go/serving/pipeline/beamio/shards.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2018 The Kythe Authors. All rights reserved.
```
Comment (schroederc) [inline kythe/go/serving/pipeline/beamio/shards.go:2]: nit: update copyright years to 2021

File: kythe/go/serving/tools/write_tables/write_tables.go [hunk 2]
```diff
@@ -64,17 +67,32 @@ var (
...
 
 func init() {
+	flag.Var(&beamInternalSharding, "beam_internal_sharding", "Internal sharding for tuning performance")
 	gsutil.Flag(&gs, "graphstore", "GraphStore to read (mutually exclusive with --entries)")
 	flag.Usage = flagutil.SimpleUsage(
 		"Creates a combined xrefs/filetree/search serving table based on a given GraphStore or stream of GraphStore-ordered entries",
 		"(--graphstore spec | --entries path) --out path")
 }
 
+type intListFlag []int
```
Comment (schroederc) [inline kythe/go/serving/tools/write_tables/write_tables.go:84]: Put this in kythe/go/util/flagutil

File: kythe/go/serving/tools/write_tables/write_tables.go [hunk 3]
```diff
@@ -64,17 +67,32 @@ var (
...
 	experimentalBeamPipeline = flag.Bool("experimental_beam_pipeline", false, "Whether to use the Beam experimental pipeline implementation")
 	beamShards               = flag.Int("beam_shards", 0, "Number of shards for beam processing. If non-positive, a reasonable default will be chosen.")
+	beamK                    = flag.Int("beam_k", 0, "Amount of memory to use when creating level DB shards.")
+	beamInternalSharding     intListFlag
 	experimentalColumnarData = flag.Bool("experimental_beam_columnar_data", false, "Whether to emit columnar data from the Beam pipeline implementation")
 	compactTable             = flag.Bool("compact_table", false, "Whether to compact the output LevelDB after its creation")
 )
 
 func init() {
+	flag.Var(&beamInternalSharding, "beam_internal_sharding", "Internal sharding for tuning performance")
```
Comment (schroederc) [inline kythe/go/serving/tools/write_tables/write_tables.go:77]: Be more specific in the flag description

File: kythe/go/serving/tools/write_tables/write_tables.go [hunk 4]
```diff
@@ -64,17 +67,32 @@ var (
 
 	experimentalBeamPipeline = flag.Bool("experimental_beam_pipeline", false, "Whether to use the Beam experimental pipeline implementation")
 	beamShards               = flag.Int("beam_shards", 0, "Number of shards for beam processing. If non-positive, a reasonable default will be chosen.")
+	beamK                    = flag.Int("beam_k", 0, "Amount of memory to use when creating level DB shards.")
```
Comment (schroederc) [inline kythe/go/serving/tools/write_tables/write_tables.go:70]: Describe units and default

File: external.bzl [hunk 5]
```diff
@@ -318,8 +318,8 @@ def _go_dependencies():
         patches = [
             "@io_kythe//third_party/go:add_export_license.patch",
         ],
-        sum = "h1:Y8Q9pZ9V8IKM8EDNOE314D6cJE0neJooK+MxBvCcs1M=",
-        version = "v2.25.0+incompatible",
+        commit = "8f219a2e16c040fe112981dffa7c993ce76e8470",
```
Comment (schroederc) [inline external.bzl:321]: The go.* files need updated to reflect this version change.

**Review by schroederc — CHANGES_REQUESTED** [review]

**Comment by creachadair** [comment]
This seems like a good idea. Are there any benchmark results that you could add to document how much this helps?

File: kythe/go/serving/tools/write_tables/write_tables.go (hunk 4 above)
Comment (johnedmonds) [reply kythe/go/serving/tools/write_tables/write_tables.go]: Updated the string. Let me know if this makes sense.

File: kythe/go/serving/tools/write_tables/write_tables.go (hunk 3 above)
Comment (johnedmonds) [reply kythe/go/serving/tools/write_tables/write_tables.go]: Updated the description. Hopefully this makes sense.

**Comment by johnedmonds** [comment]
@creachadair Thanks! Good question. I ran a few `/decorations` requests on my laptop (Mac with an SSD):
On average it took about 1 minute for a `decorations` request to localhost. I made these requests shortly after downloading the LevelDB files to my laptop. If I understand correctly, LevelDB has a thread that's sorting these shards in the background into higher levels. But last I checked, that process seems to take a few hours to untangle the overlapping shards on our database. Also since I checked, our database has grown a lot larger.

I also tested the same `/decorations` requests against our internal (remote) instance and that takes about 260ms.

So all in all, this change is about a 23000% increase in speed 🎉 over the current overlapping shards--at least for the first few hours of operation.

**Review by schroederc — APPROVED** [review]

**Pull request merged**